routing/index_graph.hpp

```diff
@@ -99,27 +106,116 @@ class IndexGraph final
 
   void GetSegmentCandidateForJoint(Segment const & parent, bool isOutgoing, std::vector<Segment> & children);
-  void ReconstructJointSegment(Segment const & parent, std::vector<Segment> const & firstChildren,
-                               std::vector<uint32_t> const & lastPointIds,
-                               bool isOutgoing, std::vector<JointEdge> & jointEdges,
-                               std::vector<RouteWeight> & parentWeights);
+  void ReconstructJointSegment(JointSegment const & parentJoint,
+                               Segment const & parent,
+                               vector<Segment> const & firstChildren,
+                               vector<uint32_t> const & lastPointIds,
+                               bool isOutgoing,
+                               vector<JointEdge> & jointEdges,
+                               vector<RouteWeight> & parentWeights,
+                               std::map<JointSegment, JointSegment> & parents);
 
   shared_ptr<Geometry> m_geometry;
   shared_ptr<EdgeEstimator> m_estimator;
   RoadIndex m_roadIndex;
   JointIndex m_jointIndex;
-  RestrictionVec m_restrictions;
+  Restrictions m_restrictionsForward;
+  Restrictions m_restrictionsBackward;
   RoadAccess m_roadAccess;
   RoutingOptions m_avoidRoutingOptions;
 };
+
+/// \note |parent| can be fake, |parentSegment| is always non-fake.
+template <typename Parent>
+bool IndexGraph::IsRestricted(Parent const & parent,
+                              uint32_t parentFeatureId,
+                              uint32_t currentFeatureId, bool isOutgoing,
+                              std::map<Parent, Parent> & parents) const
+{
+  if (parentFeatureId == currentFeatureId)
+    return false;
+
+  auto const & restrictions = isOutgoing ? m_restrictionsForward : m_restrictionsBackward;
+  auto const it = restrictions.find(currentFeatureId);
+  if (it == restrictions.cend())
+    return false;
+
+  std::vector<Parent> parentsFromCurrent;
+  auto const getParent = [&parents, &parentsFromCurrent](Parent const & p)
+  {
+    uint32_t prevFeatureId = p.GetFeatureId();
+    uint32_t curFeatureId = prevFeatureId;
+
+    auto nextParent = parents.end();
+    auto * curParrent = &p;
+    while (curFeatureId == prevFeatureId)
+    {
+      auto const parentIt = parents.find(*curParrent);
+      if (parentIt == parents.cend())
+        return false;
+
+      curFeatureId = parentIt->second.GetFeatureId();
+      nextParent = parentIt;
+      curParrent = &nextParent->second;
+    }
+
+    ASSERT(nextParent != parents.end(), ());
+    parentsFromCurrent.emplace_back(nextParent->second);
+    return true;
+  };
+
+  for (std::vector<uint32_t> const & restriction : it->second)
+  {
+    bool prevIsParent = restriction[0] == parentFeatureId;
+    if (!prevIsParent)
+      continue;
+
+    if (restriction.size() == 1)
+      return true;
+
+    // If parents empty we process only two feature restrictions
```

Comment:
`// If parents are empty we process only two feature restrictions.`?

Reply:
doen